[ALLUXIO-262] Cache incomplete blocks during file seek

peisun1115 opened the pull request:
https://alluxio.atlassian.net/browse/ALLUXIO-262

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9448/<h2>Build result: FAILURE</span></h2>[...truncated 1332 lines...][JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/s3/target/alluxio-underfs-s3-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-s3/1.1.0-SNAPSHOT/alluxio-underfs-s3-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/s3/target/alluxio-underfs-s3-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-s3/1.1.0-SNAPSHOT/alluxio-underfs-s3-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/assembly/pom.xml to org.alluxio/alluxio-assemblies/1.1.0-SNAPSHOT/alluxio-assemblies-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/pom.xml to org.alluxio/alluxio-shell/1.1.0-SNAPSHOT/alluxio-shell-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/pom.xml to org.alluxio/alluxio-underfs-oss/1.1.0-SNAPSHOT/alluxio-underfs-oss-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/target/alluxio-underfs-oss-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-oss/1.1.0-SNAPSHOT/alluxio-underfs-oss-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/target/alluxio-underfs-oss-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-oss/1.1.0-SNAPSHOT/alluxio-underfs-oss-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/client/pom.xml to org.alluxio/alluxio-keyvalue-client/1.1.0-SNAPSHOT/alluxio-keyvalue-client-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/pom.xml to org.alluxio/alluxio-underfs-hdfs/1.1.0-SNAPSHOT/alluxio-underfs-hdfs-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-hdfs/1.1.0-SNAPSHOT/alluxio-underfs-hdfs-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-hdfs/1.1.0-SNAPSHOT/alluxio-underfs-hdfs-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/pom.xml to org.alluxio/alluxio-underfs-swift/1.1.0-SNAPSHOT/alluxio-underfs-swift-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/target/alluxio-underfs-swift-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-swift/1.1.0-SNAPSHOT/alluxio-underfs-swift-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/target/alluxio-underfs-swift-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-swift/1.1.0-SNAPSHOT/alluxio-underfs-swift-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/pom.xml to org.alluxio/alluxio-examples/1.1.0-SNAPSHOT/alluxio-examples-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/pom.xml to org.alluxio/alluxio-core-server/1.1.0-SNAPSHOT/alluxio-core-server-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/pom.xml to org.alluxio/alluxio-keyvalue-common/1.1.0-SNAPSHOT/alluxio-keyvalue-common-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/pom.xml to org.alluxio/alluxio-underfs/1.1.0-SNAPSHOT/alluxio-underfs-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/pom.xml to org.alluxio/alluxio-underfs-glusterfs/1.1.0-SNAPSHOT/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-glusterfs/1.1.0-SNAPSHOT/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-glusterfs/1.1.0-SNAPSHOT/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/integration/pom.xml to org.alluxio/alluxio-integration/1.1.0-SNAPSHOT/alluxio-integration-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/pom.xml to org.alluxio/alluxio-keyvalue/1.1.0-SNAPSHOT/alluxio-keyvalue-1.1.0-SNAPSHOT.pomchannel stopped
Test FAILed.

peisun1115 commented:
IntelliJ autoformating added some changes to this PR. If you feel they are misleading for code review, I can revert them.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9449/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.alluxio:alluxio-core-client' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/9449/org.alluxio$alluxio-core-client/testReport'>org.alluxio:alluxio-core-client</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/9449/org.alluxio$alluxio-core-client/testReport/alluxio.client.file.options/InStreamOptionsTest/testEquals/'><strong>alluxio.client.file.options.InStreamOptionsTest.testEquals</strong></a></li></ul>
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9450/<h2>Build result: FAILURE</span></h2>[...truncated 1311 lines...][JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/s3/target/alluxio-underfs-s3-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-s3/1.1.0-SNAPSHOT/alluxio-underfs-s3-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/s3/target/alluxio-underfs-s3-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-s3/1.1.0-SNAPSHOT/alluxio-underfs-s3-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/assembly/pom.xml to org.alluxio/alluxio-assemblies/1.1.0-SNAPSHOT/alluxio-assemblies-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/pom.xml to org.alluxio/alluxio-shell/1.1.0-SNAPSHOT/alluxio-shell-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/pom.xml to org.alluxio/alluxio-underfs-oss/1.1.0-SNAPSHOT/alluxio-underfs-oss-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/target/alluxio-underfs-oss-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-oss/1.1.0-SNAPSHOT/alluxio-underfs-oss-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/target/alluxio-underfs-oss-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-oss/1.1.0-SNAPSHOT/alluxio-underfs-oss-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/client/pom.xml to org.alluxio/alluxio-keyvalue-client/1.1.0-SNAPSHOT/alluxio-keyvalue-client-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/pom.xml to org.alluxio/alluxio-underfs-hdfs/1.1.0-SNAPSHOT/alluxio-underfs-hdfs-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-hdfs/1.1.0-SNAPSHOT/alluxio-underfs-hdfs-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-hdfs/1.1.0-SNAPSHOT/alluxio-underfs-hdfs-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/pom.xml to org.alluxio/alluxio-underfs-swift/1.1.0-SNAPSHOT/alluxio-underfs-swift-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/target/alluxio-underfs-swift-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-swift/1.1.0-SNAPSHOT/alluxio-underfs-swift-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/target/alluxio-underfs-swift-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-swift/1.1.0-SNAPSHOT/alluxio-underfs-swift-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/pom.xml to org.alluxio/alluxio-examples/1.1.0-SNAPSHOT/alluxio-examples-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/pom.xml to org.alluxio/alluxio-core-server/1.1.0-SNAPSHOT/alluxio-core-server-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/pom.xml to org.alluxio/alluxio-keyvalue-common/1.1.0-SNAPSHOT/alluxio-keyvalue-common-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/pom.xml to org.alluxio/alluxio-underfs/1.1.0-SNAPSHOT/alluxio-underfs-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/pom.xml to org.alluxio/alluxio-underfs-glusterfs/1.1.0-SNAPSHOT/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-glusterfs/1.1.0-SNAPSHOT/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-glusterfs/1.1.0-SNAPSHOT/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/integration/pom.xml to org.alluxio/alluxio-integration/1.1.0-SNAPSHOT/alluxio-integration-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/pom.xml to org.alluxio/alluxio-keyvalue/1.1.0-SNAPSHOT/alluxio-keyvalue-1.1.0-SNAPSHOT.pomchannel stopped
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9451/
Test PASSed.

gpang commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -78,6 +78,8 @@
   /** Whether or not the current block should be cached. */
   private boolean mShouldCacheCurrentBlock;
+  /** Include incomplete blocks if Alluxio is configured to store blocks in Alluxio storage. */
+  private boolean mShouldCacheIncompleteBlock;
```
should this be final?

peisun1115 replied:
done

gpang commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -237,8 +241,9 @@ private void checkAndAdvanceBlockInStream() throws IOException {
           long blockSize = getCurrentBlockSize();
-          WorkerNetAddress address = mLocationPolicy.getWorkerForNextBlock(
-              mContext.getAlluxioBlockStore().getWorkerInfoList(), blockSize);
+          WorkerNetAddress address = mLocationPolicy
```
is this just formatting? I think I prefer the previous format.

peisun1115 replied:
Done. I use the auto-format in IntelliJ.

gpang commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -349,8 +354,10 @@ private void seekBlockInStream(long newPos) throws IOException {
           long blockSize = getCurrentBlockSize();
-          WorkerNetAddress address = mLocationPolicy.getWorkerForNextBlock(
-              mContext.getAlluxioBlockStore().getWorkerInfoList(), blockSize);
+          WorkerNetAddress address = mLocationPolicy
```
is this just formatting? I think I prefer the previous format.

peisun1115 replied:
done.

gpang commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -404,4 +411,64 @@ private void updateBlockInStream(long blockId) throws IOException {
+   *            pos <= mFileLength)
+   * @throws IOException if the seek fails due to an error accessing the stream at the position
+   */
+  private void seekWithoutCachingIncompleteBlocks(long pos) throws IOException {
```
This can probably just be called `seekInternal()`, and the other version can be `seekInternalWithCaching()`? What do you think?

peisun1115 replied:
done

gpang commented on core/client/src/main/java/alluxio/client/file/options/InStreamOptions.java:
```diff
@@ -32,6 +32,8 @@
 public final class InStreamOptions {
   private FileWriteLocationPolicy mLocationPolicy;
   private ReadType mReadType;
+  /** Include incomplete blocks if Alluxio is configured to store blocks in Alluxio storage. */
```
This comment is confusing. Instead of "include" maybe use "cache".

peisun1115 replied:
done

gpang commented on core/client/src/main/java/alluxio/client/file/options/InStreamOptions.java:
```diff
@@ -88,6 +91,23 @@ public InStreamOptions setReadType(ReadType readType) {
+  /**
+   * @return true if incomplete block caching is enabled
+   */
+  public boolean shouldCacheIncompleteBlock() {
```
this should be named like a getter, so `getCacheIncompleteBlock()`

peisun1115 replied:
done

gpang commented on core/client/src/main/java/alluxio/client/file/options/InStreamOptions.java:
```diff
@@ -88,6 +91,23 @@ public InStreamOptions setReadType(ReadType readType) {
+   *
+   * @return the updated options object
+   */
+  public InStreamOptions enableCacheIncompleteBlock() {
```
this should be named like a setter, so `setCacheIncompleteBlock(boolean cacheIncompleteBlock)`

peisun1115 replied:
done

apc999 replied:
Same comment I left. it seems the code is not really changed. (did you manage to push the local change?)

peisun1115 replied:
I am not sure what you meant. Let me know if you still see the problem. Thanks

gpang commented on core/client/src/test/java/alluxio/client/file/FileInStreamTest.java:
```diff
@@ -257,6 +257,59 @@ public void testSeek() throws IOException {
+        mCacheStreams.get(1).getWrittenData());
+
+    // Seek forward within block 1 many times.
+    for (int i = 0; i <= seekAmount / 2; ++i) {
```
The style in Alluxio is to use `i++`

peisun1115 replied:
done

gpang commented on core/client/src/test/java/alluxio/client/file/FileInStreamTest.java:
```diff
@@ -257,6 +257,59 @@ public void testSeek() throws IOException {
+        mCacheStreams.get(1).getWrittenData());
+
+    // Seek backward within block 1 many times.
+    for (int i = seekAmount / 2; i >= 0; --i) {
```
The style in Alluxio is to use `i++`

peisun1115 replied:
done

gpang commented on core/common/src/main/resources/alluxio-default.properties:
```diff
@@ -142,6 +142,7 @@ alluxio.user.failed.space.request.limits=3
 alluxio.user.file.buffer.bytes=1MB
 alluxio.user.file.master.client.threads=10
 alluxio.user.file.readtype.default=CACHE_PROMOTE
+alluxio.user.file.cache.incomplete.block=false
```
This parameter should be added to the documentation.

peisun1115 replied:
done

gpang commented:
Thanks @peisun1115 for this feature! I left some comments.

calvinjia commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -404,4 +411,64 @@ private void updateBlockInStream(long blockId) throws IOException {
+  }
+
+  /**
+   * Seeks to a file position. Blocks are cached even if they are fully read. This is only called by
```
If they are not fully read?

peisun1115 replied:
done

calvinjia commented:
@peisun1115 is the intended behavior consistent with the ticket? It seems like using this option will not cache if a block if the read is just of a small portion of the block. Also if the user seeks to a position inside a large file, all the blocks before that position will need to be cached which is not the intention right?

apc999 commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -78,6 +78,8 @@
   private boolean mClosed;
   /** Whether or not the current block should be cached. */
   private boolean mShouldCacheCurrentBlock;
+  /** Include incomplete blocks if Alluxio is configured to store blocks in Alluxio storage. */
```
the javadoc is a bit unclear. do you mean:

Whether to store incomplete blocks in Alluxio ?

peisun1115 replied:
done

apc999 commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -195,12 +198,13 @@ public void seek(long pos) throws IOException {
-    seekBlockInStream(pos);
-    checkAndAdvanceBlockInStream();
-    mCurrentBlockInStream.seek(mPos % mBlockSize);
+    Preconditions
```
NIT: I like the previous formating better :(

peisun1115 replied:
done.  I used IntelliJ's autoformat

apc999 commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -404,4 +411,64 @@ private void updateBlockInStream(long blockId) throws IOException {
+  }
+
+  /**
+   * Read the current block till pos or the end of the current block if some conditions are met.
```
NIT: Reads

peisun1115 replied:
done

apc999 commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -195,12 +198,13 @@ public void seek(long pos) throws IOException {
-    mCurrentBlockInStream.seek(mPos % mBlockSize);
+    Preconditions
+        .checkArgument(pos < mFileLength, PreconditionMessage.ERR_SEEK_PAST_END_OF_FILE, pos);
+    if (mShouldCacheIncompleteBlock) {
```
what if `mShouldCacheCurrentBlock` is `false`? do we still call `seekWithCachingIncompleteBlocks` in this case?

peisun1115 replied:
yes, that is expected.  I think it is correctly handled now.

apc999 commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -404,4 +411,64 @@ private void updateBlockInStream(long blockId) throws IOException {
+   */
+  private void seekWithCachingIncompleteBlocks(long pos) throws IOException {
+    boolean isInCurrentBlock = pos / mBlockSize == mPos / mBlockSize;
+    if (mShouldCacheCurrentBlock) {
```
I feel this method needs some work to be more clear and easy to reason.

it looks to me several dimensions to make it complicated with multiple cases:
(1) whether this seek is forward or backward
(2) whether the seek is in the same block or not
(3) `mShouldCacheCurrentBlock` is true or not.

Could we have a cleaner branching here so we don't mistakenly get some corner case wrong?

peisun1115 replied:
I agree with you. I actually started with the way you are suggesting, but the code will be much longer and have some code duplication. The current logic is this:
1.  Cache the current block or read till pos depending on whether pos > mPos or pos is out of the current block.
2.  If pos is in current block, set mPos to pos. Otherwise, jump to pos and read the prefix of the new block. 

Basically, it is trying to maintain an invariant that we discussed in person.

apc999 commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -404,4 +411,64 @@ private void updateBlockInStream(long blockId) throws IOException {
+   * @param pos at most read till pos
+   * @throws IOException if read or cache write fails
+   */
+  private void readCurrentBlockTill(long pos) throws IOException {
```
is it the case `pos < mPos` this is a noop?

peisun1115 replied:
yes. I added a comment

apc999 commented on core/client/src/test/java/alluxio/client/file/FileInStreamTest.java:
```diff
@@ -257,6 +257,59 @@ public void testSeek() throws IOException {
+   * Tests seek with incomplete block caching enabled. It seeks within a block forward
+   * and backward, across blocks.
+   *
+   * @throws IOException when reading from the stream fails
```
you don't need to javadoc the exceptions for test cases :)

peisun1115 replied:
Done. I did that because other tests in this file are doing that.

apc999 commented on core/client/src/test/java/alluxio/client/file/FileInStreamTest.java:
```diff
@@ -257,6 +257,59 @@ public void testSeek() throws IOException {
+   * @throws IOException when reading from the stream fails
+   */
+  @Test
+  public void testSeekWithCachingIncompleteBlocks() throws IOException {
```
alluxio naming convention: `seekWithCachingIncompleteBlocksTest`

peisun1115 replied:
done

peisun1115 commented:
@gpang @apc999 @calvinjia 

I explained this PR and compared it with the other approach (prefetch every block) in https://docs.google.com/document/d/1mR1_NPt60smz5nWMAyeUZfXNllrKKmc39CaUORb1oPg/edit

I can change this PR to do prefetching if you guys prefer.  I will address the comments once we agree on the approach.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9490/
Test PASSed.

peisun1115 commented:
@calvinjia 's question:
""" is the intended behavior consistent with the ticket? It seems like using this option will not cache if a block if the read is just of a small portion of the block. Also if the user seeks to a position inside a large file, all the blocks before that position will need to be cached which is not the intention right?""""

Yes, I think the behavior is consistent with the ticket in terms of the intended behavior but not in terms of the method used.  A block will be cached if the read is just a small portion of the block. If the user seeks to a pos inside a large file, only the first and last block (where the cursor "started, stopped") will be cached.

@gpang  @apc999  I've addressed your comments.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9519/
Test PASSed.

peisun1115 commented:
@apc999  @gpang  @calvinjia 

I have verified this PR by reading parquet files from spark.  PTAL?

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9548/
Test PASSed.

apc999 commented on docs/_data/table/en/user-configuration.yml:
```diff
@@ -33,6 +33,8 @@ alluxio.user.file.writetype.default:
   `CACHE_THROUGH` (try to cache, write to UnderFS synchronously), `THROUGH` (no cache, write to
   UnderFS synchronously).
+alluxio.user.file.cache.incomplete.block:
+  A block is cached to Alluxio even if the block is not fully read when this flag is enabled.
```
NIT (try to be more explicit):

When read type is `CACHE_PROMOTE` or `CACHE` and this property is set to true, the entire block will be cached by Alluxio space even if the client only reads a part of this block.

apc999 replied:
please also update the Chinese version accordingly.

peisun1115 replied:
done

apc999 commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -404,4 +412,66 @@ private void updateBlockInStream(long blockId) throws IOException {
+
+  /**
+   * Reads the current block till pos or the end of the current block if some conditions are met.
+   * No-op is pos <= mPos.
```
NIT (try to be more explicit):
Reads till the file offset (mPos) equals pos or the end of the current block (whichever is achieved) if pos > mPos; otherwise a no-op.

peisun1115 replied:
done

apc999 commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -404,4 +412,66 @@ private void updateBlockInStream(long blockId) throws IOException {
+   * Reads the current block till pos or the end of the current block if some conditions are met.
+   * No-op is pos <= mPos.
+   *
+   * @param pos at most read till pos
```
pos is in terms of the file offset, right? better to be clear here.

peisun1115 replied:
done

apc999 commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -404,4 +412,66 @@ private void updateBlockInStream(long blockId) throws IOException {
+  }
+
+  /**
+   * Seeks to a file position. Blocks are cached if they are fully read. This is only called by
```
"if they are fully read" or "even if they are not fully read"?

peisun1115 replied:
done

apc999 commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -195,12 +201,13 @@ public void seek(long pos) throws IOException {
-    mCurrentBlockInStream.seek(mPos % mBlockSize);
+    Preconditions.checkArgument(
+        pos < mFileLength, PreconditionMessage.ERR_SEEK_PAST_END_OF_FILE, pos);
+    if (!mShouldCacheIncompleteBlock) {
```
```java
if (mShouldCacheCurrentBlock && mShouldCacheIncompleteBlock) {
  mCurrentBlockInStream.seek(mPos % mBlockSize);
  seekInternalWithCachingIncompleteBlock(pos);
} else {
  seekBlockInStream(pos);
  seekInternal(pos);
  checkAndAdvanceBlockInStream();
} 
```

is it?

peisun1115 replied:
Discussed in person

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9549/
Test PASSed.

apc999 commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -404,4 +412,66 @@ private void updateBlockInStream(long blockId) throws IOException {
+   * @throws IOException if the seek fails due to an error accessing the stream at the position
+   */
+  private void seekInternalWithCachingIncompleteBlock(long pos) throws IOException {
+    boolean isInCurrentBlock = pos / mBlockSize == mPos / mBlockSize;
```
Alternative approach to code the logic:
```java
boolean isInCurrentBlock = pos / mBlockSize == mPos / mBlockSize;
if (pos < mPos) {
  // backward, but still in the same block
  if (isInCurrentBlock) {
    readCurrentBlockTill(mFileLength);
    mCurrentBlockInStream.seek(mPos % mBlockSize);
  } else {
   // Seek to some block before
   readCurrentBlockTill(mFileLength);
   seekBlockInStream(pos / mBlockSize * mBlockSize);
   checkAndAdvanceBlockInStream();
   readCurrentBlockTill(pos);
  } 
} else {
  // forward, but still in the same block
  if (isInCurrentBlock) {
    readCurrentBlockTill(pos);
  } else {
   // Seek to some block after
   readCurrentBlockTill(mFileLength);
   seekBlockInStream(pos / mBlockSize * mBlockSize);
   checkAndAdvanceBlockInStream();
   readCurrentBlockTill(pos);
  } 
}
```

peisun1115 replied:
Discussed in person.

apc999 commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -404,4 +412,66 @@ private void updateBlockInStream(long blockId) throws IOException {
+      // Cache the current block or read forward to pos.
+      readCurrentBlockTill(pos > mPos ? pos : mFileLength);
+    }
+    if (mPos == pos) {
```
as discussed, move this check down. and put more comments.

peisun1115 replied:
Done. I've added  more comments to explain this.

peisun1115 commented:
@apc999  Thanks a lot for the review. I have addressed your comments.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9550/
Test PASSed.

apc999 commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -404,4 +412,77 @@ private void updateBlockInStream(long blockId) throws IOException {
+    // If seeks within the current block, directly seeks to pos if we are not yet there.
+    // If seeks outside the current block, seek to the beginning of that block first, then
+    // cache the prefix (pos % mBlockSize) of that block.
+
```
NIT: unnecessary empty line

peisun1115 replied:
done

apc999 commented on docs/_data/table/cn/user-configuration.yml:
```diff
@@ -25,6 +25,8 @@ alluxio.user.file.readtype.default:
   创建Alluxio文件时的默认写类型。可选值为`MUST_CACHE` (数据仅仅存储在Alluxio中，并且必须存储在其中),
   `CACHE_THROUGH` (尽量缓冲数据，同时同步写入到底层文件系统), `THROUGH` (不缓冲数据，同步写入到底层文件系统)。
+alluxio.user.file.cache.incomplete.block:
+  当读取类型是 CACHE 或者 CACHE_PROMOTE 的时候, 如果设置这个变量为真, 没有完全读取的数据块也会被存到Alluxio内.
```
NIT: should we quote `CACHE` and `CACHE_PROMOTE` like <pre>\`CACHE\`</pre>  here?

peisun1115 replied:
done

apc999 commented on core/client/src/main/java/alluxio/client/file/options/InStreamOptions.java:
```diff
@@ -88,6 +91,23 @@ public InStreamOptions setReadType(ReadType readType) {
+   *
+   * @return the updated options object
+   */
+  public InStreamOptions setCacheIncompleteBlock() {
```
it seems in the code base, the convention for setters of a boolean field in options takes one param, like
`public LoadMetadataOptions setRecursive(boolean recursive)`

peisun1115 replied:
done

apc999 commented on core/client/src/main/java/alluxio/client/file/options/InStreamOptions.java:
```diff
@@ -88,6 +91,23 @@ public InStreamOptions setReadType(ReadType readType) {
+  }
+
+  /**
+   * Enable incomplete block caching.
```
NIT: Enables

peisun1115 replied:
done

apc999 commented on core/client/src/test/java/alluxio/client/file/FileInStreamTest.java:
```diff
@@ -257,6 +257,57 @@ public void testSeek() throws IOException {
+   * and backward, across blocks.
+   */
+  @Test
+  public void seekWithCachingIncompleteBlocksTest() throws IOException {
```
one suggestion for the test: IMO, it is more readable / maintainble if we break the big chunk of test case into multiple simpler test cases. For example in your case, I would suggest to have at least four test cases covering {forward,backword}x{withinBlock, crossBlock} four basic scenarios as we discussed offline.

Then based on that, we could have more complicated tests for different seek sequences.

Sorry for pushing on this but I think it is a fairly important and tricky code change. To get better confidence, we should probably have more tests. What do you think?

peisun1115 replied:
Thanks for the suggestion. Done.

peisun1115 commented:
@apc999  Thanks for the review. I have addressed your comments. PTAL

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9558/<h2>Build result: FAILURE</span></h2>[...truncated 1330 lines...][JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/s3/target/alluxio-underfs-s3-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-s3/1.1.0-SNAPSHOT/alluxio-underfs-s3-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/s3/target/alluxio-underfs-s3-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-s3/1.1.0-SNAPSHOT/alluxio-underfs-s3-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/assembly/pom.xml to org.alluxio/alluxio-assemblies/1.1.0-SNAPSHOT/alluxio-assemblies-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/pom.xml to org.alluxio/alluxio-shell/1.1.0-SNAPSHOT/alluxio-shell-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/pom.xml to org.alluxio/alluxio-underfs-oss/1.1.0-SNAPSHOT/alluxio-underfs-oss-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/target/alluxio-underfs-oss-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-oss/1.1.0-SNAPSHOT/alluxio-underfs-oss-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/target/alluxio-underfs-oss-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-oss/1.1.0-SNAPSHOT/alluxio-underfs-oss-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/client/pom.xml to org.alluxio/alluxio-keyvalue-client/1.1.0-SNAPSHOT/alluxio-keyvalue-client-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/pom.xml to org.alluxio/alluxio-underfs-hdfs/1.1.0-SNAPSHOT/alluxio-underfs-hdfs-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-hdfs/1.1.0-SNAPSHOT/alluxio-underfs-hdfs-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-hdfs/1.1.0-SNAPSHOT/alluxio-underfs-hdfs-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/pom.xml to org.alluxio/alluxio-underfs-swift/1.1.0-SNAPSHOT/alluxio-underfs-swift-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/target/alluxio-underfs-swift-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-swift/1.1.0-SNAPSHOT/alluxio-underfs-swift-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/target/alluxio-underfs-swift-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-swift/1.1.0-SNAPSHOT/alluxio-underfs-swift-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/pom.xml to org.alluxio/alluxio-examples/1.1.0-SNAPSHOT/alluxio-examples-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/pom.xml to org.alluxio/alluxio-core-server/1.1.0-SNAPSHOT/alluxio-core-server-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/pom.xml to org.alluxio/alluxio-keyvalue-common/1.1.0-SNAPSHOT/alluxio-keyvalue-common-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/pom.xml to org.alluxio/alluxio-underfs/1.1.0-SNAPSHOT/alluxio-underfs-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/pom.xml to org.alluxio/alluxio-underfs-glusterfs/1.1.0-SNAPSHOT/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-glusterfs/1.1.0-SNAPSHOT/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-glusterfs/1.1.0-SNAPSHOT/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/integration/pom.xml to org.alluxio/alluxio-integration/1.1.0-SNAPSHOT/alluxio-integration-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/pom.xml to org.alluxio/alluxio-keyvalue/1.1.0-SNAPSHOT/alluxio-keyvalue-1.1.0-SNAPSHOT.pomchannel stopped
Test FAILed.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9561/
Test PASSed.

apc999 commented:
LGTM. Please rebase your PR with latest master merged. thanks !

calvinjia commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -404,4 +412,75 @@ private void updateBlockInStream(long blockId) throws IOException {
+    Preconditions.checkState(len <= Integer.MAX_VALUE);
+
+    byte[] buffer = new byte[(int) len];
+    read(buffer);
```
Since we are discarding the bytes anyway, does it make sense to allocate such a large buffer?

peisun1115 replied:
good point. Done

calvinjia commented on conf/alluxio-site.properties.template:
```diff
@@ -75,6 +75,7 @@ alluxio.user.failed.space.request.limits=3
 alluxio.user.file.buffer.bytes=1MB
 alluxio.user.file.master.client.threads=10
 alluxio.user.file.readtype.default=CACHE_PROMOTE
+alluxio.user.file.cache.incomplete.block=false
```
Not sure if this might be confusing, since we have a concept of `complete` and `incomplete` files.

peisun1115 replied:
I feel this is fine since complete file is not exposed to client. I am open to change this if you have another suggestion.

gpang replied:
I agree with @calvinjia . A complete (or incomplete) block means something in Alluxio.

instead of "incomplete", maybe "partialread"?

calvinjia commented on core/client/src/main/java/alluxio/client/file/options/InStreamOptions.java:
```diff
@@ -88,6 +91,24 @@ public InStreamOptions setReadType(ReadType readType) {
+  /**
+   * @return true if incomplete block caching is enabled
+   */
+  public boolean getCacheIncompleteBlock() {
```
I think we use `is...` for getting booleans

peisun1115 replied:
Done.

calvinjia commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -404,4 +412,75 @@ private void updateBlockInStream(long blockId) throws IOException {
+    if (len <= 0) {
+      return;
+    }
+    Preconditions.checkState(len <= Integer.MAX_VALUE);
```
This implicitly caps block sizes at `Integer.MAX_VALUE`.

peisun1115 replied:
done

calvinjia commented:
@peisun1115 
In the future, would it be possible to refactor this logic into a class which extends `FileInStream` instead of putting the logic in the class?

peisun1115 commented:
Calvin, I have addressed your comments.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9609/
Test PASSed.

gpang commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -80,7 +80,9 @@
   protected boolean mShouldCacheCurrentBlock;
-  /** Current position of the stream. */
+  /** Include incomplete blocks if Alluxio is configured to store blocks in Alluxio storage. */
+  private boolean mShouldCacheIncompleteBlock;
```
maybe Incomplete -> PartialRead

also, how does this variable interact with `mShouldCacheCurrentBlock`?

peisun1115 replied:
done.

mShouldCacheCurrentBlock is per block and keep changing, mShouldCacheIncompleteBlock is set when the InStream object is created.

gpang commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -228,12 +234,13 @@ public void seek(long pos) throws IOException {
-    checkAndAdvanceBlockInStream();
-    mCurrentBlockInStream.seek(mPos % mBlockSize);
+    Preconditions.checkArgument(
+        pos < mFileLength, PreconditionMessage.ERR_SEEK_PAST_END_OF_FILE, pos);
```
use `validPosition()`

peisun1115 replied:
I think it is better to keep it as it is in this PR to keep the error msg consistent with the seek() in BufferedBlockInStream.java .

gpang replied:
I don't understand. The existing code uses `validPosition(pos)`, but this PR changed the condition. Why is that?

peisun1115 replied:
Just to explain what I meant: 
validPosition only checks pos < mFileLength, we still need to keep the other check for pos >= 0.  we generate ERR_SEEK_PAST_END_OF_FILE error msg for validPosition check failure which assumes that validPositions checking pos < mFileLength.  So I think if we use validPosition, it is better to change validPostion to include pos >=0 and update the error msgs.

but I don't really have strong opinions on this. I have changed it back.

gpang commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -345,7 +352,7 @@ private void checkAndAdvanceBlockInStream() throws IOException {
           WorkerNetAddress address = mLocationPolicy.getWorkerForNextBlock(
               mContext.getAlluxioBlockStore().getWorkerInfoList(), getBlockSizeAllocation(mPos));
-          // Don't cache the block to somewhere that already has it.
+         // Don't cache the block to somewhere that already has it.
```
I think this should be changed back?

peisun1115 replied:
done

gpang commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -508,4 +515,76 @@ private void updateBlockInStream(long blockId) throws IOException {
+
+  /**
+   * Seeks to a file position. Blocks are cached even if they are not fully read. This is only
+   * called by {@link FileInStream#seek}.
```
What happens on a `skip()`? Shouldn't skip behave similarly to seek?

peisun1115 replied:
no. I am not changing skip().  No one is using it for now.  We can change it later when it is used.

gpang commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -508,4 +515,76 @@ private void updateBlockInStream(long blockId) throws IOException {
+   * @param pos file offset
+   * @throws IOException if read or cache write fails
+   */
+  private void readCurrentBlockTill(long pos) throws IOException {
```
I don't think I like "till". Maybe `readCurrentBlockToPos` or something?

peisun1115 replied:
done

gpang commented:
@peisun1115 Thanks! I left a few comments.

peisun1115 commented:
@gpang  I have addressed your comments.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9651/<h2>Build result: FAILURE</span></h2>[...truncated 1310 lines...][JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/s3/target/alluxio-underfs-s3-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-s3/1.1.0-SNAPSHOT/alluxio-underfs-s3-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/s3/target/alluxio-underfs-s3-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-s3/1.1.0-SNAPSHOT/alluxio-underfs-s3-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/assembly/pom.xml to org.alluxio/alluxio-assemblies/1.1.0-SNAPSHOT/alluxio-assemblies-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/pom.xml to org.alluxio/alluxio-shell/1.1.0-SNAPSHOT/alluxio-shell-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/pom.xml to org.alluxio/alluxio-underfs-oss/1.1.0-SNAPSHOT/alluxio-underfs-oss-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/target/alluxio-underfs-oss-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-oss/1.1.0-SNAPSHOT/alluxio-underfs-oss-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/target/alluxio-underfs-oss-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-oss/1.1.0-SNAPSHOT/alluxio-underfs-oss-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/client/pom.xml to org.alluxio/alluxio-keyvalue-client/1.1.0-SNAPSHOT/alluxio-keyvalue-client-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/pom.xml to org.alluxio/alluxio-underfs-hdfs/1.1.0-SNAPSHOT/alluxio-underfs-hdfs-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-hdfs/1.1.0-SNAPSHOT/alluxio-underfs-hdfs-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-hdfs/1.1.0-SNAPSHOT/alluxio-underfs-hdfs-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/pom.xml to org.alluxio/alluxio-underfs-swift/1.1.0-SNAPSHOT/alluxio-underfs-swift-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/target/alluxio-underfs-swift-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-swift/1.1.0-SNAPSHOT/alluxio-underfs-swift-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/target/alluxio-underfs-swift-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-swift/1.1.0-SNAPSHOT/alluxio-underfs-swift-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/pom.xml to org.alluxio/alluxio-examples/1.1.0-SNAPSHOT/alluxio-examples-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/pom.xml to org.alluxio/alluxio-core-server/1.1.0-SNAPSHOT/alluxio-core-server-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/pom.xml to org.alluxio/alluxio-keyvalue-common/1.1.0-SNAPSHOT/alluxio-keyvalue-common-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/pom.xml to org.alluxio/alluxio-underfs/1.1.0-SNAPSHOT/alluxio-underfs-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/pom.xml to org.alluxio/alluxio-underfs-glusterfs/1.1.0-SNAPSHOT/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-glusterfs/1.1.0-SNAPSHOT/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-glusterfs/1.1.0-SNAPSHOT/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/integration/pom.xml to org.alluxio/alluxio-integration/1.1.0-SNAPSHOT/alluxio-integration-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/pom.xml to org.alluxio/alluxio-keyvalue/1.1.0-SNAPSHOT/alluxio-keyvalue-1.1.0-SNAPSHOT.pomchannel stopped
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9652/
Test PASSed.

gpang commented:
Thanks @peisun1115 , I just had one last question.

peisun1115 commented:
@gpang fixed

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9657/
Test PASSed.

apc999 commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -508,4 +515,82 @@ private void updateBlockInStream(long blockId) throws IOException {
+      if (mPos == pos) {
+        return;
+      }
+
```
if we reach here, that means seeks backwards, right?
if yes, please javadoc this.

peisun1115 replied:
done. FYI, I am refactoring this file to make it clearer. I feel FileInStream is a little hard to follow and easy to make mistakes when I make changes to it.

apc999 commented on core/client/src/main/java/alluxio/client/file/FileInStream.java:
```diff
@@ -508,4 +515,82 @@ private void updateBlockInStream(long blockId) throws IOException {
+        readCurrentBlockToPos(pos);
+      } else {
+        seekBlockInStream(pos);
+        mCurrentBlockInStream.seek(mPos % mBlockSize);
```
commenting that `mPos == pos` here?

peisun1115 replied:
done

apc999 commented:
I left two post-LGTM comments, if fixed/answered, LGTM to me

peisun1115 commented:
@apc999  Fixed. Thank you.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9675/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.alluxio:alluxio-tests' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/9675/org.alluxio$alluxio-tests/testReport'>org.alluxio:alluxio-tests</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/9675/org.alluxio$alluxio-tests/testReport/alluxio.collections/IndexedSetConcurrencyTest/basicConcurrentUpdateTest/'><strong>alluxio.collections.IndexedSetConcurrencyTest.basicConcurrentUpdateTest</strong></a></li></ul>
Test FAILed.

peisun1115 commented:
Jerkins, test this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9678/
Test PASSed.

gpang commented:
LGTM

calvinjia commented:
LGTM
